src/libsyntax/parse/parser.rs

```diff
@@ -3690,6 +3690,40 @@ impl<'a> Parser<'a> {
         }))
     }
 
+    /// A wrapper around `parse_pat` with some special error handling
+    /// for the top-level pattern in a `let` statement or `for` loop.
+    pub fn parse_binding_pat(&mut self) -> PResult<'a, P<Pat>> {
+        let pat = self.parse_pat()?;
+        if self.token == token::Comma {
+            // An unexpected comma after an ident pattern is a clue that the
+            // user (perhaps more accustomed to some other language) forgot the
+            // parentheses in what should have been a tuple pattern; emit a
+            // suggestion-enhanced error here rather than choking on the comma
+            // later.
+            if let PatKind::Ident(_, _, ref sub) = pat.node {
+                if sub.is_none() { // just `ident`, not `ident @ pattern`
```

Comment:
`PatKind::Ident` and `sub.is_none` here are artificial restrictions.
What we want is to parse `PAT, PAT, ..., PAT` instead of just `PAT` and return `PatKind::Tuple` containing those patterns.
It's not very important if those pattern are identifiers or not - if we can easily recover more, we should do it.
(You may want to modify `parse_pat_tuple_elements` to return partial result even in case of failure.)